Sources/ApodiniDeployBuildSupport/Options/OptionKey.swift

```diff
@@ -0,0 +1,63 @@
+//
+//  OptionKey.swift
+//
+//
+//  Created by Lukas Kollmer on 2021-02-06.
+//
+
+import Foundation
+
+
+public class AnyOptionKey<OuterNS: OuterNamespace>: Codable, Hashable, Equatable, CustomStringConvertible {
```

Comment:
A more general question: We have the Options that are defined as part of ApodiniDeploy and we have the Options that are part of the main Apodini Target. Wouldn't it make sense to move the general construct in its own target, e.g. ApodiniOptions and import this in Apodini and Apodini Deploy so we don't have two options implementations laying around? I think this would also become beneficial when we start thinking about the metadata DSL in the future. Or would there be a specific reason for keeping the two implementations separate?

In addition this might also be beneficial for the work that @eldcn is doing in the future. The current WebService Model already contains a lot of important information and maybe adding some more general context and information to that would be good as this would allow us to e.g. reuse the WebService model for the diffing of web service interfaces too.

Reply:
Yea there's definitely a lot of duplication between the two Options APIs.

I did look into simply using @Supereg's implementation, but there are a couple of limitations to my use case which the other implementation doesn't cover.

I did look into using @Supereg's implementation, but there's a couple of constraints which apply to my use case, which do not apply to the existing implementation in the Apodini target.
The main issue is that my options API needs to be 100% codable: since the options are instantiated in one process (the web service) and then used in another (the deployment provider), we have to somehow transfer them across that barrier w/out any loss in type safety.
It would probably be possible to rewrite the existing options API in a way to make it Codable, but that would come at the cost of a worse API usage experience.
Example: The existing API currently uses `ObjectIdentifier` to compare option keys, meaning you can define option keys ([eg the `.http` option](https://github.com/Apodini/Apodini/blob/0653bb4f07dd67bbe45ebffb3e6b7d07a16f6903/Sources/Apodini/Properties/Options/HTTPParameterMode.swift)) simply by defining a static property. (This works because the `.http` object being used when specifying an `@Parameter` option is then the same one being used later to read that option.)
In my case, because these two objects (the one used when specifying the option in the web service, and the one used to read the option in the deployment provider) are completely distinct objects, using a static property and comparing the object's identity simply would not work, and i instead had to add string keys to my `AnyOptionKey` type: its still strongly typed when you use it locally and when you de/encode it, but the actual comparison of option keys has to fall back to the untyped strings used in the background.

Obviously we could just add the concept of explicitly defined keys to the existing options implementation, but that would result in an objectively worse API: you can already define options without having to key them, so why require adding keys if they're not actually required for the in-process usage of the options api?

There's a couple of other issues which are specific to my implementation, mainly the concept of unresolved options (the conditional ones), which may or may not contribute to the full set of options, depending on a handler-based condition. I'm not really sure if/how this could be best integrated into the existing API.



The options API used for `@Parameters`

I also didn't want to

Reply:
Thanks for the detailed overview! As mentioned in a comment above maybe the other way around would work? So instead of using the Options API that we have at the moment for the Parameters we could use the more generalized API that you have at the moment using the inner and outer namespaces. I could even imagine that the conditional API would be a very nice addition in the future to enable the usage of conditionals e.g. in the Metadata DSL. And if the Parameters Options and all other Options follow the same API it would me even simpler to move some of the aspects into the Metadata DSL in the future.